kani-compiler/src/kani_middle/resolve.rs

```diff
@@ -6,426 +6,363 @@
                 Some((hir_id, _)) => tcx.hir().local_def_id(hir_id),
             };
-            return Some(Path::new(
-                Base::LocalModule { id: crate_root, may_be_external_path },
-                segments,
-            ));
+            Ok(Path { base: crate_root.to_def_id(), segments: segments.collect() })
         }
         SELF => {
-            segments.pop_front();
+            segments.next();
+            resolve_super(tcx, current_module, segments)
+        }
+        _ => {
+            let path = resolve_super(tcx, current_module, segments)?;
+            if !path.segments.is_empty() {
+                let next_name = path.segments.first().unwrap();
+                let def_id = resolve_in_module(tcx, path.base, &next_name)?;
+                Ok(Path { base: def_id, segments: Vec::from(&path.segments[1..]) })
+            } else {
+                Ok(path)
+            }
         }
-        _ => (),
     }
+}
 
-    // Pop up the module stack until we account for all the `super` prefixes.
+/// Pop up the module stack until we account for all the `super` prefixes.
+/// This method will error out if it tries to backtrace from the root crate.
+fn resolve_super<I>(
+    tcx: TyCtxt,
+    current_module: LocalDefId,
+    mut segments: Peekable<I>,
+) -> Result<Path, ResolveError>
+where
+    I: Iterator<Item = String>,
+{
     let current_module_hir_id = tcx.hir().local_def_id_to_hir_id(current_module);
     let mut parents = tcx.hir().parent_iter(current_module_hir_id);
     let mut base_module = current_module;
-    while segments.front().map(String::as_str) == Some(SUPER) {
-        segments.pop_front();
-        let parent = parents.next().map(|p| p.0).or_else(|| {
-            tracing::debug!("Unable to normalize path `{name}`: too many `super` qualifiers");
-            None
-        })?;
-        base_module = tcx.hir().local_def_id(parent);
+    while segments.next_if(|segment| segment == SUPER).is_some() {
+        if let Some((parent, _)) = parents.next() {
+            debug!("parent: {parent:?}");
+            base_module = tcx.hir().local_def_id(parent);
+        } else {
+            return Err(ResolveError {
+                msg: "there are too many leading `super` keywords".to_string(),
+            });
+        }
     }
-
-    Some(Path::new(Base::LocalModule { id: base_module, may_be_external_path }, segments))
+    debug!("base: {base_module:?}");
+    Ok(Path { base: base_module.to_def_id(), segments: segments.collect() })
 }
 
-/// Resolves an external path.
-fn resolve_external(tcx: TyCtxt, mut segments: Segments) -> Option<DefId> {
-    tracing::debug!("Resolving `{}` in the external prelude", segments_to_string(&segments));
-    let first = segments.pop_front().or_else(|| {
-        tracing::debug!("Unable to resolve the empty path");
-        None
-    })?;
-    for crate_num in tcx.crates(()) {
+/// Resolves an external crate name.
+fn resolve_external(tcx: TyCtxt, name: &str) -> Option<DefId> {
```

Comment:
I was expecting this to return `Result<DefId, ...>`. Is there any reason to keep `Option<DefId>`?

Reply:
Yes. I've been trying to use `Option<>` when the only possible outcomes are item was found or item was not found. I've been using result when the outcome can be also be an error.

Reply:
Only brought this up because the usage in `resolve_prefix` would benefit from a `Result<T>` type. I don't feel too strong about this one, to be honest.

Reply:
Gotcha... but I don't think it would work so well in `resolve_relative`. So if you don't mind I will just keep it as an `Option<>`.

Reply:
Sounds good, thanks for the explanation!